File: src/uniquify.c
#include "ast.h"
#include "defs.h"
#include "table.h"
#include <string.h>

char *rewrite_var(char *p, int num) {
  int len = strlen(p);
  char *buf = malloc(len + 2);
  strcpy(buf, p);
  buf[len] = num + '0';
  buf[len + 1] = '\0';
  free((void *)p);
  return buf;
}

void uniquify2(ASTNode *node, Table *t) {
  int cnt;
  int len;
  char *buf;
  switch (node->token) {
  case Program:
    uniquify2(node->lhs, t);
    break;
  case Neg:
    uniquify2(node->lhs, t);
    break;
  case Add:
    uniquify2(node->lhs, t);
    uniquify2(node->rhs, t);
    break;
  case Var:
    cnt = table_get(t, (char *)node->value);
    node->value = (size_t)rewrite_var((char *)node->value, cnt);
    break;
  case Let:
    cnt = table_get(t, (char *)node->lhs->value);
    // increase suffix
    table_store(t, (char *)node->lhs->value, cnt + 1);
    uniquify2(node->rhs, t);
    // set back suffix
    table_store(t, (char *)node->lhs->value, cnt);
    node->lhs->value = (size_t)rewrite_var((char *)node->lhs->value, cnt + 1);
    break;
  default:
    break;
  }
}

void uniquify(ASTNode *node) {
  Table t;
  table_init(&t);
  uniquify2(node, &t);
}

File: src/pass/uniquify.rs
use crate::ast::Node;
use std::collections::HashMap;

fn rewrite_var(mut name: String, cnt: usize) -> String {
    name.push_str(&format!("_{}", cnt));
    name
}

fn uniquify_inner(node: Box<Node>, cxt: &mut HashMap<String, usize>) -> Box<Node> {
    use Node::*;

    let node = match *node {
        Program(sub_node) => Program(uniquify_inner(sub_node, cxt)),
        Neg(sub_node) => Neg(uniquify_inner(sub_node, cxt)),
        Add(lhs, rhs) => Add(uniquify_inner(lhs, cxt), uniquify_inner(rhs, cxt)),
        Var(var_name) => {
            let count = cxt.get(&var_name).map(|cnt| *cnt).unwrap_or_default();
            let new_var_name = rewrite_var(var_name, count);
            Var(new_var_name)
        }
        Let { name, value, exp } => {
            let count = cxt.get(&name).map(|cnt| *cnt).unwrap_or_default() + 1;
            // increase cnt in sub node
            cxt.insert(name.clone(), count);
            let sub_node = uniquify_inner(exp, cxt);
            // set cnt back
            cxt.insert(name.clone(), count - 1);
            let new_var_name = rewrite_var(name, count);
            Let {
                name: new_var_name,
                value,
                exp: sub_node,
            }
        }
        _ => *node,
    };
    Box::new(node)
}

pub fn uniquify(node: Box<Node>) -> Box<Node> {
    let mut cxt = HashMap::default();
    uniquify_inner(node, &mut cxt)
}
